### PR TITLE
More efficient way from "linux/circ_buf.h"

### PR DESCRIPTION
**Pull request by R32**

### CODE REVIEW

**Comment by markknol** [comment]
Looks good to me!

@Simn @Aurel300 Can you do sanity check on this?

File: assets/content/cookbook/Data structures/ring-array.md [hunk 1]
```diff
@@ -5,82 +5,78 @@
 A fixed ring array are especially useful when you need a hard upper bound for how much data can be in the queue.
```
Comment (Aurel300) [inline assets/content/cookbook/Data structures/ring-array.md]: > ... *is* especially useful ...

File: assets/content/cookbook/Data structures/ring-array.md [hunk 2]
```diff
@@ -5,82 +5,78 @@
...
-
-  var i: Int;
-  var start: Int;
-  var left: Int;
+  var head: Int;
+  var tail: Int;
+  var cap: Int;
   var a: haxe.ds.Vector<T>;
 
+  // NOTE: len must be power of 2
```
Comment (Aurel300) [inline assets/content/cookbook/Data structures/ring-array.md]: Maybe would be best if this was checked for and `throw`n?

File: assets/content/cookbook/Data structures/ring-array.md [hunk 3]
```diff
@@ -5,82 +5,78 @@
...
-      ++ j;
-    }
+
+  public function toString() {
+    return '[head: $head, tail: $tail, capacity: $cap]';
   }
-  public inline function toString() {
-    return '[i: $i, start: $start, len: $len, left: $left]';
+
+  public inline function CIRC_CNT() return (head - tail) & cap;
```
Comment (Aurel300) [inline assets/content/cookbook/Data structures/ring-array.md]: I understand the `CIRC_*` methods are taken from Linux, but you should change their names. At least `count` (`CIRC_CNT`) and `space` (`CIRC_SPACE`) can be properties with an inline getter.

File: assets/content/cookbook/Data structures/ring-array.md [hunk 4]
```diff
@@ -5,82 +5,78 @@
...
+
+  public inline function CIRC_SPACE() return (tail - head - 1) & cap;
+
+  public inline function CIRC_CNT_TO_END() {
```
Comment (Aurel300) [inline assets/content/cookbook/Data structures/ring-array.md]: Neither `CIRC_CNT_TO_END` nor `CIRC_SPACE_TO_END` is used – remove them, or demonstrate their usage somehow?

File: assets/content/cookbook/Data structures/ring-array.md [hunk 5]
```diff
@@ -113,7 +109,7 @@ It's easy to implement `undo/redo` operations.
 
 class Main {
   static function main() {
-    var h = new History<Int>(3);
+    var h = new History<Int>(4);
     h.add(1);
     h.add(2);
     h.add(3);
```
Comment (Aurel300) [inline assets/content/cookbook/Data structures/ring-array.md]: Can't comment on the line below (L116) – `overrided` is not a word, maybe it's best to say `overrides the 1`.

**Comment by R32** [comment]
[Rendered version](https://github.com/R32/code-cookbook/blob/patch-1/assets/content/cookbook/Data%20structures/ring-array.md)

**Comment by markknol** [comment]
Thanks, I've just merged it!

**Pull request merged**